File: sdk/typescript/src/context.ts
import type { SdkClient } from './client.js';
import {
  EventType,
  SyncMode,
  UserFilterMode,
  type Document,
  type DocumentMetadata,
  type DocumentPermissions,
  type ConnectorEventPayload,
  type GroupMembershipEventPayload,
} from './models.js';
import { ContentStorage } from './storage.js';
import { getLogger } from './logger.js';

const logger = getLogger('sdk:context');

/** Buffer thresholds (size, timeMs) per sync mode. `null` timeMs = flush-on-emit. */
function thresholdsFor(syncMode: SyncMode): { size: number; timeMs: number | null } {
  if (syncMode === SyncMode.FULL) return { size: 500, timeMs: 300_000 };
  if (syncMode === SyncMode.REALTIME) return { size: 1, timeMs: null };
  return { size: 100, timeMs: 60_000 }; // Incremental (default)
}

export class SyncContext {
  private readonly client: SdkClient;
  private readonly _syncRunId: string;
  private readonly _sourceId: string;
  private _state: Record<string, unknown>;
  private readonly abortController: AbortController;
  private _documentsEmitted = 0;
  private _documentsScanned = 0;
  private readonly _contentStorage: ContentStorage;
  private readonly _syncMode: SyncMode;
  private readonly _isResume: boolean;
  private readonly bufferSizeThreshold: number;
  private readonly bufferTimeThresholdMs: number | null;
  private eventBuffer: ConnectorEventPayload[] = [];
  private oldestEventAt: number | null = null;
  private readonly _sourceType: string | null;
  private readonly _userFilterMode: UserFilterMode;
  private readonly _userWhitelist: ReadonlySet<string>;
  private readonly _userBlacklist: ReadonlySet<string>;

  constructor(
    client: SdkClient,
    syncRunId: string,
    sourceId: string,
    state?: Record<string, unknown>,
    syncMode: SyncMode = SyncMode.INCREMENTAL,
    documentsScanned = 0,
    documentsUpdated = 0,
    options?: {
      isResume?: boolean;
      sourceType?: string | null;
      userFilterMode?: UserFilterMode;
      userWhitelist?: string[] | null;
      userBlacklist?: string[] | null;
    }
  ) {
    this.client = client;
    this._syncRunId = syncRunId;
    this._sourceId = sourceId;
    this._state = state ?? {};
    this.abortController = new AbortController();
    this._contentStorage = new ContentStorage(client, syncRunId);
    this._syncMode = syncMode;
    this._isResume = options?.isResume ?? false;
    const thresholds = thresholdsFor(syncMode);
    this.bufferSizeThreshold = thresholds.size;
    this.bufferTimeThresholdMs = thresholds.timeMs;
    // Seed counters from the dispatch payload so resume continues from
    // the manager's running tally rather than restarting at zero.
    this._documentsScanned = documentsScanned;
    this._documentsEmitted = documentsUpdated;
    this._sourceType = options?.sourceType ?? null;
    this._userFilterMode = options?.userFilterMode ?? UserFilterMode.ALL;
    // Store as lowercased set so shouldIndexUser is case-insensitive
    // (mirrors Python's should_index_user; deliberately differs from
    // Rust's case-sensitive Source::should_index_user).
    this._userWhitelist = new Set(
      (options?.userWhitelist ?? []).map((e) => e.toLowerCase())
    );
    this._userBlacklist = new Set(
      (options?.userBlacklist ?? []).map((e) => e.toLowerCase())
    );
  }

  get syncRunId(): string {
    return this._syncRunId;
  }

  get sourceId(): string {
    return this._sourceId;
  }

  get state(): Record<string, unknown> {
    return this._state;
  }

  get contentStorage(): ContentStorage {
    return this._contentStorage;
  }

  get documentsEmitted(): number {
    return this._documentsEmitted;
  }

  get documentsScanned(): number {
    return this._documentsScanned;
  }

  /**
   * The sync mode the manager dispatched this run with — full, incremental,
   * or realtime. Connectors should branch reset/resume behavior on this,
   * not on `state` presence: a manual "Full" trigger from the UI carries
   * existing state but should still reset cursors and run delete
   * reconciliation.
   */
  get syncMode(): SyncMode {
    return this._syncMode;
  }

  get isResume(): boolean {
    return this._isResume;
  }

  get sourceType(): string | null {
    return this._sourceType;
  }

  private async bufferEvent(event: ConnectorEventPayload): Promise<void> {
    this.eventBuffer.push(event);
    if (this.oldestEventAt === null) {
      this.oldestEventAt = Date.now();
    }

    const sizeHit = this.eventBuffer.length >= this.bufferSizeThreshold;
    const timeHit =
      this.bufferTimeThresholdMs !== null &&
      this.oldestEventAt !== null &&
      Date.now() - this.oldestEventAt >= this.bufferTimeThresholdMs;
    if (sizeHit || timeHit) {
      await this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.eventBuffer.length === 0) {
      return;
    }
    const batch = this.eventBuffer;
    this.eventBuffer = [];
    this.oldestEventAt = null;
    await this.client.emitEventBatch(this._syncRunId, this._sourceId, batch);
  }

  async emit(doc: Document): Promise<void> {
    // Shim Document.title into metadata.title — the indexer reads only
    // metadata.title and falls back to "Untitled" otherwise. Non-mutating
    // spread so the caller's Document is never modified (broader than
    // Python's conditional shim, which skips bare-metadata docs).
    const metadata: DocumentMetadata = { ...(doc.metadata ?? {}) };
    if (!metadata.title) {
      metadata.title = doc.title;
    }
    const event: ConnectorEventPayload = {
      type: EventType.DOCUMENT_CREATED,
      sync_run_id: this._syncRunId,
      source_id: this._sourceId,
      document_id: doc.external_id,
      content_id: doc.content_id,
      metadata,
      permissions: doc.permissions,
      attributes: doc.attributes,
    };
    await this.bufferEvent(event);
    this._documentsEmitted++;
  }

  async emitUpdated(doc: Document): Promise<void> {
    const metadata: DocumentMetadata = { ...(doc.metadata ?? {}) };
    if (!metadata.title) {
      metadata.title = doc.title;
    }
    const event: ConnectorEventPayload = {
      type: EventType.DOCUMENT_UPDATED,
      sync_run_id: this._syncRunId,
      source_id: this._sourceId,
      document_id: doc.external_id,
      content_id: doc.content_id,
      metadata,
      permissions: doc.permissions,
      attributes: doc.attributes,
    };
    await this.bufferEvent(event);
    this._documentsEmitted++;
  }

  async emitDeleted(externalId: string): Promise<void> {
    const event: ConnectorEventPayload = {
      type: EventType.DOCUMENT_DELETED,
      sync_run_id: this._syncRunId,
      source_id: this._sourceId,
      document_id: externalId,
    };
    await this.bufferEvent(event);
  }

  async emitGroupMembership(
    groupEmail: string,
    memberEmails: string[],
    groupName?: string,
  ): Promise<void> {
    const event: GroupMembershipEventPayload = {
      type: EventType.GROUP_MEMBERSHIP_SYNC,
      sync_run_id: this._syncRunId,
      source_id: this._sourceId,
      group_email: groupEmail,
      group_name: groupName,
      member_emails: memberEmails,
    };
    await this.bufferEvent(event);
  }

  emitError(externalId: string, error: string): void {
    logger.warn(`Document error for ${externalId}: ${error}`);
  }

  async incrementScanned(): Promise<void> {
    this._documentsScanned++;
    await this.client.incrementScanned(this._syncRunId);
  }

  /**
   * Bump the server-side documents_updated counter for this sync run.
   *
   * Mirrors Rust SDK's increment_updated. The connector author chooses
   * what counts as "updated" — typically called once per successfully
   * persisted emit/emitUpdated, or batched (e.g. once per processed
   * chunk). Errors propagate so the caller can decide whether a stat
   * miss should fail the sync or be swallowed.
   */
  async incrementUpdated(count: number): Promise<void> {
    if (count <= 0) return;
    await this.client.incrementUpdated(this._syncRunId, count);
  }

  /**
   * Check whether a user should be indexed under this source's user-filter
   * settings. Mirrors the Python SDK's should_index_user.
   *
   * Connectors call this themselves before emitting per-user records — the
   * SDK does NOT auto-skip on emit, since the connector knows whether a
   * given doc is "owned by a user" or workspace-shared.
   *
   * Empty email: ALL admits, WHITELIST/BLACKLIST reject (matches Python).
   * Comparison is case-insensitive (Rust's equivalent is case-sensitive —
   * deliberate divergence; lowercasing here is the friendlier default and
   * matches operator expectations).
   */
  shouldIndexUser(userEmail: string): boolean {
    if (this._userFilterMode === UserFilterMode.ALL) return true;
    const email = userEmail.toLowerCase();
    if (!email) return false;
    if (this._userFilterMode === UserFilterMode.WHITELIST) {
      return this._userWhitelist.has(email);
    }
    if (this._userFilterMode === UserFilterMode.BLACKLIST) {
      return !this._userBlacklist.has(email);
    }
    return true;
  }

  /**
   * Checkpoint state for resumability. Call periodically for long syncs.
   *
   * Flushes buffered events first — without this, a crash right after
   * checkpointing would lose events that the connector considered emitted
   * (the next run resumes past them).
   */
  async saveCheckpoint(checkpoint: Record<string, unknown>): Promise<void> {
    await this.flush();
    this._state = checkpoint;
    await this.client.updateCheckpoint(this._syncRunId, checkpoint);
    await this.client.heartbeat(this._syncRunId);
  }

  async saveState(state: Record<string, unknown>): Promise<void> {
    await this.saveCheckpoint(state);
  }

  async complete(newState?: Record<string, unknown>): Promise<void> {
    await this.flush();
    if (newState !== undefined) {
      this._state = newState;
      await this.client.updateCheckpoint(this._syncRunId, newState);
    }
    await this.client.complete(this._syncRunId);
  }

  async fail(error: string): Promise<void> {
    try {
      await this.flush();
    } catch (e) {
      logger.warn(
        `flush before fail() failed (continuing): sync_run=${this._syncRunId}: ${e}`
      );
    }
    await this.client.fail(this._syncRunId, error);
  }

  isCancelled(): boolean {
    return this.abortController.signal.aborted;
  }

  _setCancelled(): void {
    this.abortController.abort();
  }
}

File: sdk/rust/src/context.rs
use crate::client::SdkClient;
use anyhow::Result;
use shared::models::{ConnectorEvent, SourceType, SyncType};
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use tracing::warn;

#[derive(Clone)]
pub struct SyncContext {
    sdk_client: SdkClient,
    sync_run_id: String,
    source_id: String,
    source_type: SourceType,
    sync_mode: SyncType,
    is_resume: bool,
    cancelled: Arc<AtomicBool>,
}

impl SyncContext {
    pub fn new(
        sdk_client: SdkClient,
        sync_run_id: String,
        source_id: String,
        source_type: SourceType,
        sync_mode: SyncType,
        cancelled: Arc<AtomicBool>,
    ) -> Self {
        Self::new_with_resume(
            sdk_client,
            sync_run_id,
            source_id,
            source_type,
            sync_mode,
            false,
            cancelled,
        )
    }

    pub fn new_with_resume(
        sdk_client: SdkClient,
        sync_run_id: String,
        source_id: String,
        source_type: SourceType,
        sync_mode: SyncType,
        is_resume: bool,
        cancelled: Arc<AtomicBool>,
    ) -> Self {
        Self {
            sdk_client,
            sync_run_id,
            source_id,
            source_type,
            sync_mode,
            is_resume,
            cancelled,
        }
    }

    pub fn sdk_client(&self) -> &SdkClient {
        &self.sdk_client
    }

    pub fn sync_run_id(&self) -> &str {
        &self.sync_run_id
    }

    pub fn source_id(&self) -> &str {
        &self.source_id
    }

    pub fn source_type(&self) -> SourceType {
        self.source_type
    }

    pub fn sync_mode(&self) -> SyncType {
        self.sync_mode
    }

    pub fn is_resume(&self) -> bool {
        self.is_resume
    }

    pub fn is_cancelled(&self) -> bool {
        self.cancelled.load(Ordering::SeqCst)
    }

    /// Emit a single event. Events are buffered in memory and auto-flushed
    /// according to the sync's mode (see [`thresholds_for`]):
    /// - Full: 500 events or 5min, whichever first
    /// - Incremental: 100 events or 60s
    /// - Realtime: flush on every emit
    ///
    /// If an auto-flush fails, the error is returned to the caller so the
    /// connector knows the event was not persisted before checkpointing.
    pub async fn emit_event(&self, event: ConnectorEvent) -> Result<()> {
        self.sdk_client
            .emit_event(&self.sync_run_id, &self.source_id, event)
            .await?;
        Ok(())
    }

    /// Flush all buffered events for this (sync_run_id, source_id) pair.
    pub async fn flush(&self) -> Result<()> {
        self.sdk_client
            .flush_events(&self.sync_run_id, &self.source_id)
            .await?;
        Ok(())
    }

    /// Flush all buffered events across all (sync_run_id, source_id) pairs.
    pub async fn flush_all(&self) -> Result<()> {
        self.sdk_client.flush_all().await?;
        Ok(())
    }

    pub async fn extract_and_store_content(
        &self,
        data: Vec<u8>,
        mime_type: &str,
        filename: Option<&str>,
    ) -> Result<String> {
        Ok(self
            .sdk_client
            .extract_and_store_content(&self.sync_run_id, data, mime_type, filename)
            .await?)
    }

    pub async fn store_content(&self, content: &str) -> Result<String> {
        Ok(self
            .sdk_client
            .store_content(&self.sync_run_id, content)
            .await?)
    }

    pub async fn increment_scanned(&self, count: i32) -> Result<()> {
        self.sdk_client
            .increment_scanned(&self.sync_run_id, count)
            .await?;
        Ok(())
    }

    pub async fn increment_updated(&self, count: i32) -> Result<()> {
        self.sdk_client
            .increment_updated(&self.sync_run_id, count)
            .await?;
        Ok(())
    }

    /// Mark sync as completed. Flushes any buffered events first so the
    /// completion never races ahead of the final events for this sync.
    /// Status flip only — counts come from `increment_scanned`/`updated`,
    /// checkpoint from `save_checkpoint`.
    pub async fn complete(&self) -> Result<()> {
        self.sdk_client.complete(&self.sync_run_id).await?;
        Ok(())
    }

    /// Mark sync as failed. Best-effort flush of buffered events first — if
    /// the flush itself fails we log and proceed, because marking the sync as
    /// failed is more important than preserving partial progress.
    pub async fn fail(&self, error: &str) -> Result<()> {
        if let Err(e) = self.flush_all().await {
            warn!(
                "SDK: flush before fail() failed (continuing): sync_run={}: {}",
                self.sync_run_id, e
            );
        }
        self.sdk_client.fail(&self.sync_run_id, error).await?;
        Ok(())
    }

    pub async fn heartbeat(&self) -> Result<()> {
        self.sdk_client.heartbeat(&self.sync_run_id).await?;
        Ok(())
    }

    pub async fn cancel(&self) -> Result<()> {
        self.sdk_client.cancel(&self.sync_run_id).await?;
        Ok(())
    }

    /// Checkpoint state for resumability. Flushes buffered events first —
    /// without this, a crash after checkpointing would lose events that the
    /// connector considered emitted (the next run resumes past them).
    pub async fn save_checkpoint(&self, checkpoint: serde_json::Value) -> Result<()> {
        self.sdk_client
            .save_checkpoint(&self.sync_run_id, &self.source_id, checkpoint)
            .await?;
        Ok(())
    }

    #[deprecated(note = "use save_checkpoint")]
    pub async fn save_connector_state(&self, state: serde_json::Value) -> Result<()> {
        self.save_checkpoint(state).await
    }

    pub async fn get_user_email_for_source(&self) -> Result<String> {
        Ok(self
            .sdk_client
            .get_user_email_for_source(&self.source_id)
            .await?)
    }
}
